File: Data Structures & Algorithms/regular-expression-matching/submission-8.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        N,M =len(s),len(p)
        def dfs(i,j):
            if j==M:
                return i == N
            m = i<N and (s[i] == p[j] or p[j] == ".")
            if j+1<M and p[j+1] == "*":
                return (m and dfs(i+1,j)) or dfs(i,j+2)
            elif m:
                return dfs(i+1,j+1)
            else:
                return False
        return dfs(0,0)




        # nnn .*
        # i.   
        # j
        


        

File: Data Structures & Algorithms/search-for-word-ii/submission-8.java
public class Node{
    Node[] children = new Node[26];
    boolean end = false;
}



class Solution {
    Node root = new Node();
    Set<String> result = new HashSet<>();
    boolean[][] visit;
    int rows,cols;
    public List<String> findWords(char[][] board, String[] words) {
        rows = board.length;
        cols = board[0].length;
        visit = new boolean[rows][cols];
        for(String w:words){
            var cur = root;
            for(char c:w.toCharArray()){
                var index = c-'a';
                if(cur.children[index]==null){
                    cur.children[index] = new Node();
                }
                cur = cur.children[index];
            }
            cur.end = true;
        }
        for(int i = 0;i<rows;i++){
            for(int j = 0;j<cols;j++){
                dfs(i,j,root,"",board);
            }
        }
        return new ArrayList<>(result);
        
    }

    public void dfs(int i, int j, Node node, String word, char[][] board){
        if(i<0 || j<0|| i==rows||j==cols||visit[i][j]||node.children[board[i][j]-'a'] == null){
            return;
        }
        visit[i][j] = true;
        node = node.children[board[i][j]-'a'];
        word += board[i][j];
        if(node.end){
            result.add(word);
        }
        dfs(i+1,j,node,word,board);
        dfs(i,j+1,node,word,board);
        dfs(i-1,j,node,word,board);
        dfs(i,j-1,node,word,board);
        visit[i][j] = false;
    }
}
